packages/eslint-plugin/src/rules/no-magic-numbers.ts

```diff
@@ -70,6 +73,15 @@ export default util.createRule<Options, MessageIds>({
           return;
         }
 
+        // Check if the node is a type index
+        if (
+          options.ignoreTypeIndexes &&
+          (typeof node.value === 'number' || typeof node.value === 'bigint') &&
+          isAncestorTSIndexedAccessType(node)
+        ) {
+          return;
```

Comment:
FWIW, this looks good to me.
The only difference with my implementation is that I reported invalid nodes in this statement, similar to "Check if the node is a readonly class property". This way, we can return early which should be faster?

Also, good catch for the union types - this was something that I overlooked.

Reply:
@timdeschryver Thanks! I refactored things a little in d1e65bec86620ad4f3631316334605507920b531 so if any of the four ignorable TypeScript rules are triggered it will report early. Is that better?

Reply:
Yea, this should work.
I would remove the `else if (isAllowed === false)` statement to remove the indentation, and perhaps move the `isAllowed` logic into its own method.

Keep in mind that I'm not a member of typescript-eslint 😅

Reply:
@timdeschryver Realized I never responded to this, sorry - and thank you for the feedback.

> remove the `else if (isAllowed === false)` statement

I don’t think I can remove the `isAllowed === false` case because it’s possible for `isAllowed` to be `undefined`, so I intend to explicitly check if it is `true` or `false` in that `if/else if` block. If it is neither it will fall back to the base rule.

Perhaps I should move the final `rules.Literal(node);` line of the function into an `else` block to make that more explicit?

> move the `isAllowed` logic into its own method

Looking at a few other rule definitions at random it does seem like they keep the core rule logic in the `create` function, which makes sense to me. I’m not sure moving `isAllowed` into its own function would improve readability, and feels like it would simply cause people to have to jump to look elsewhere for what is essentially the meat of the rule.

Feel free to push back if you (or @JoshuaKGoldberg?) disagree though, and I will happily take another look.

Reply:
No problem @davecardwell !
Thanks for the elaboration, that does make sense.